CIP-0188? | Token-Owned Addresses (#1200)
cardano-foundation/CIPs: CIP-0188/README.md

---
CIP: ?
Title: Token-Owned Addresses

> Category: Tokens

Reviewer: ```suggestion
Category: Tools
```
To confirm at next CIP meeting after any initial discussion here (unless agreed immediately), based on a broad implementation beyond tokens themselves: specifies script interaction, recommends wallet behaviour, and demands interpretation by indexers & explorers.  Leaving GitHub `Category` tag off until this question is at least temporarily settled at the meeting in 3 days' time).

Author: I can see the case for Tools, but my initial lean is still Tokens.
The closest precedents seem to be CIP-0025 and CIP-0068: both are categorized under Tokens while also defining off-chain metadata structure, wallet/viewer interpretation, and third-party lookup or retrieval behaviour.
TOA’s primary subject is native-token/NFT interoperability  (deterministic addresses controlled by a Cardano asset class). The wallet/indexer/explorer sections are included to make that token interoperability consistent, not because the proposal is primarily a tooling API.

That said, I’m happy to change it to Tools if that is considered the better fit.

Reviewer: @mariusgeorgescu that's a good analogy.  I'll still bring it up at the meeting — if anyone thinks some advantage could be obtained by marking this as `Tools` instead — although historically that hasn't led to better review since nobody from the analytics quarter (and very few from the wallets quarter) has habitually engaged with the CIP process so far: so we would get better feedback based on the `Tokens` category & I'll set that category on GitHub in the meantime & resolve this.

- `toa-bench` — a synthetic-context micro-benchmark for the validator with baseline-diff support for regression checks on execution units and script size.

The off-chain library, HTTP API, generator, and benchmark are non-normative. Conformance is determined by the validator artifact and the normative test vectors, per *Reference Artifacts — Normative components*.

#### Generated bindings (supplementary) — external

Supplementary language bindings MAY be generated from the externally-published `toa.tx3` reference file or produced on demand; all such bindings are non-normative.

#### Demo frontend — external

A non-normative reference frontend exercising address derivation, deposit, spend, and inspection is hosted at [toa.e7d.tech](https://toa.e7d.tech). The frontend consumes the HTTP API exposed by `interaction-api` in the reference repository and is provided for illustrative purposes only — conformance is determined by the validator artifact and the normative test vectors, not by the frontend.

> ### Test Vector Format

Reviewer: Would this not change at some point & therefore be better to link to the hash & other deep details in the `en7angled/toa` repository?  We can't always rely on the CIP process itself to ensure this is updated — more attention would be paid to the consistency of that material in the source repository, and so whatever mutable details possible should be linked to an official (secure & accountable) source where it would be more maintainable.

The other CIP editors (particuarly @Ryun1) will be helpful in agreeing whether the bulk of the testing parameters can be included in the external repository, or perhaps are defined in the right amount of detail here in the CIP... although with @Ryun1's own CIPs it's proven helpful to put the test vectors & related material in separate file(s).

In any case, I can see @mariusgeorgescu you've already considered this & we can discuss this further before deciding in consensus.

Author: Updated in cd93314 and en7angled/toa@0.1.1.

Took the conformance-vs-documentation cut: anything that defines a conformance gate stays inline in the CIP, and anything that's schema documentation (intrinsically mutable, evolves with the JSON shape) moved out.

**Inline in the CIP:**
- The (a) / (b) conformance split with their MUST rules.
- The Coverage MUSTs list (what scenarios test vectors must cover).
- The pointer to the external file with its blake2b-256 content hash.

**Moved to `test-vectors/README.md` in the reference repository:**
- The JSON envelope shape example.
- Top-level and per-vector field semantics.
- How to regenerate the vectors via `toa-gen-vectors`.

Both files (`toa-v1.json` and the new `README.md`) are content-hashed in the CIP, so silent drift is detectable. CIP Test Vector Format section dropped from ~70 to ~25 lines.

cc @Ryun1 — would appreciate a sanity check on this split.

Reviewer: Confirming fixed in latest version (and probably long ago).

  - **Inline Datum** — any non-unit inline datum. MAY carry application-specific payloads (offer markers, inventory metadata, badges, receipts, state markers). Wallets that do not recognise the schema MUST treat the deposit as opaque value plus an unknown datum payload; they MUST NOT assume the datum is semantically meaningful unless they understand the schema.
  - **Datum Hash** — a bare 32-byte hash with no on-chain preimage. Spending requires the off-chain datum preimage.
  - **No Datum** — no datum attached.

  The TOA v1 validator does not reject a UTxO based on its datum classification.
- **`toa_version`:** an integer baked into the validator parameters that pins the standard revision. A change in `toa_version` yields a different address for the same `(policy_id, asset_name)`.

### Canonical Parameter Encoding (CDDL)

```cddl
toa_params_v1 = #6.121([
    toa_version : uint,                ; = 1 in v1
    policy_id   : bytes .size 28,      ; blake2b-224 of minting script
    asset_name  : bytes .size (0..32), ; raw asset name bytes
])
```

> `#6.121` is the CBOR tag for `Constr 0` in PlutusData. Serialisation MUST be canonical PlutusData CBOR as produced by the Plutus V3 `serialiseData` builtin: a `Constr 0` with a non-empty field list — always the case for `toa_params_v1`'s three fields — is encoded with an *indefinite-length array* (`0x9f` … `0xff`), while an empty `Constr 0 []` encodes as the definite-length `d8 79 80` (the Unit Datum byte sequence in *Definitions*); integers and bytestrings use smallest CBOR major-type form. For the unsigned-integer field this means: values `0..23` are encoded in the initial byte alone (major type 0 — `toa_version = 1` → `0x01`, `toa_version = 2` → `0x02`); values `24..255` as `0x18` followed by one value byte. This is byte-identical to what `Codec.Serialise.serialise` produces over `PlutusLedgerApi.V3.Data` in `plutus-core`. Note this is **not** strict RFC 8949 §4.2.1 (which mandates definite-length items); see *Mandatory Normative Artifacts — Identity of the parameter encoder*.

Reviewer: ```suggestion
`#6.121` is the CBOR tag for `Constr 0` in PlutusData. Serialisation MUST be canonical PlutusData CBOR as produced by the Plutus V3 `serialiseData` builtin: a `Constr 0` with a non-empty field list — always the case for `toa_params_v1`'s three fields — is encoded with an *indefinite-length array* (`0x9f` … `0xff`), while an empty `Constr 0 []` encodes as the definite-length `d8 79 80` (the Unit Datum byte sequence in *Definitions*); integers and bytestrings use smallest CBOR major-type form. For the unsigned-integer field this means: values `0..23` are encoded in the initial byte alone (major type 0 — `toa_version = 1` → `0x01`, `toa_version = 2` → `0x02`); values `24..255` as `0x18` followed by one value byte. This is byte-identical to what `Codec.Serialise.serialise` produces over `PlutusLedgerApi.V3.Data` in `plutus-core`. Note this is **not** strict RFC 8949 §4.2.1 (which mandates definite-length items); see [Mandatory Normative Artifacts](#mandatory-normative-artifacts) — *Identity of the parameter encoder*.
```

```
1. params_cbor   := serialiseData(TOAParamsV1(toa_version = 1, policy_id, asset_name))
2. chunked_param := consByteString(lengthOfByteString(params_cbor), params_cbor)
                      || consByteString(0x00, "")
3. flat_body     := FLAT_PREFIX_TOA_V1 || chunked_param || FLAT_SUFFIX_TOA_V1
4. script_bytes  := cbor_bytestring_header(lengthOfByteString(flat_body)) || flat_body
5. hash          := blake2b_224(0x03 || script_bytes)        // 0x03 = Plutus V3 language tag
6. cred          := payment_credential(script_hash = hash)
7. address       := enterprise_address(network, cred)        // no stake credential in v1
8. bech32        := encode_bech32(prefix in {addr, addr_test}, address)
```

Each ingredient:

- `params_cbor` is the canonical PlutusData CBOR as produced by the Plutus V3 `serialiseData` builtin (`Constr 0` fields encoded with indefinite-length array `0x9f…0xff`; integers and bytestrings in smallest CBOR major-type form). The CDDL `toa_params_v1` in *Canonical Parameter Encoding* defines the input.

> - `FLAT_PREFIX_TOA_V1` (528 bytes) and `FLAT_SUFFIX_TOA_V1` (1 byte = `0x01`) are normative byte-level artifacts pinned by blake2b-256 — see *Mandatory Normative Artifacts*.

Reviewer: ```suggestion
- `FLAT_PREFIX_TOA_V1` (528 bytes) and `FLAT_SUFFIX_TOA_V1` (1 byte = `0x01`) are normative byte-level artifacts pinned by blake2b-256 — see [Mandatory Normative Artifacts](#mandatory-normative-artifacts).
```

6. **Publish at least one non-normative reference off-chain implementation**, including address derivation, deposit, spend, and NFT carry-through. (Done — `offchain-lib` + `interaction-api` + [toa.e7d.tech](https://toa.e7d.tech).)
7. **Verify against independent implementation #1** — an external party (wallet team, library author, or auditor) reproduces every test vector byte-for-byte for address derivation and pass/fail-correct for validator scenarios.
8. **Verify against independent implementation #2** — a second independent party does the same against a different stack (e.g. lucid-evolution, MeshJS, cardano-transaction-lib, or PyCardano).
9. **Integrate explorer and wallet display** — at least one Cardano explorer (Cexplorer / Pool.pm / Cardanoscan / equivalent) and at least one mainstream wallet label TOA addresses as such and surface the policy classification (`ProvenClosed` / `KnownOpen` / `Unknown`) and self-deposit state.

> Steps 1–6 are owned by the CIP authors. Steps 7–9 are owned by the ecosystem and are the primary gating condition for activation.

Reviewer: I noticed that the tick-box format (which we generally prefer) which had been employed throughout the _Path to Active_ as of https://github.com/cardano-foundation/CIPs/pull/1200#discussion_r3293480225 has been thrown out.
* This would be a lot easier for editors & future reviewers to read than writing `Done` somewhere in these long items but I know you've seen your document both ways and we would honour your preference in doing it without the tickboxes.
* I've heard there are rendering problems (on certain browsers, believe it or not) with tickboxes in ordered lists: another reason to give you license in this matter because we couldn't vouch that they'd appear properly everywhere... and likely the indication of a sequential process is important so I wouldn't suggest reverting to an unordered list.
* Maybe you could highlight the word **Done** if you don't think it would disrupt the flow or appearance of the text: to make it easier for readers to follow the completion state.

I do think this final item must be modified for accuracy.  It seemed clearer before when steps 1-6 were implicit with merging the document, but like we said in https://github.com/cardano-foundation/CIPs/pull/1200#discussion_r3293481246 it is commonly known that CIP editors have to verify statements in documents before merge... and we don't want it to sound like editors are developing the spec & reference implementation 😅

```suggestion
CIP editors will verify completion of each step and apply the `Active` tag when fully complete.  Steps 7–9 are owned by the ecosystem and are the primary criteria for activation.
```

- [CIP-0160](https://cips.cardano.org/cip/CIP-0160) (Receiving Script Purpose and Addresses) proposes a ledger-level mechanism — a new `Receiving` script purpose plus a `ProtectedAddress` concept — that would let scripts validate UTxO creation at their address. TOA v1 does not depend on CIP-0160 and remains compatible with the current unprotected script-address model; if CIP-0160 is adopted, a future TOA version may use it to restrict or validate deposits into TOAs (see *Open Questions and Limitations*).

**Within the Aiken / Plutus design-patterns ecosystem:**

> - The [`parameter-validation`](https://github.com/Anastasia-Labs/aiken-design-patterns/blob/main/lib/aiken-design-patterns/parameter-validation.ak) module of Anastasia Labs' [aiken-design-patterns](https://github.com/Anastasia-Labs/aiken-design-patterns) library captures, in a more general form, the byte-level technique TOA's R uses to reconstruct a parameterised script's hash without invoking `apply_params` at runtime — splicing precomputed prefix and suffix bytes around the parameter region and hashing the result with `blake2b_224`. Example usage of the library is at [`validators/examples/parameter-validation.ak`](https://github.com/Anastasia-Labs/aiken-design-patterns/blob/main/validators/examples/parameter-validation.ak). TOA's R is conceptually a specific instance of the same pattern, specialised to a single composite parameter and pinned to a single canonical artifact. The encoding choices differ: the library hashes each parameter to a constant 28-byte footprint (`blake2b_224(serialise(param))`) so that arbitrary parameter types — and, in particular, multiple interleaved parameters — share a uniform on-chain layout; TOA embeds the raw `serialiseData` output of a single composite parameter `Constr 0 [toa_version, policy_id, asset_name]`, variable in length, framed by a fixed 1-byte suffix. The two encodings reflect different design goals — generality across parameter shapes versus a specialised single-parameter form — not a quality judgement. See *Why raw parameter embedding rather than hashing parameters?* in *Rationale — Design Decisions* for the specific reasons TOA chose the latter.

Reviewer 2: Good feedback, I will make sure we add variants of the library functions which provide direct argument support such that when the caller is certain that the argument they are parameterizing the contract with is byte aligned already (as is the case for `Constr` BuiltinData) then no hashing is required, likewise if the argument is a currency symbol, script hash or other hashed already hashed or byte aligned parameter.

Reviewer 3: @colll78 @mariusgeorgescu this is now merged and included in [`v1.8.0`](https://github.com/Anastasia-Labs/aiken-design-patterns/releases/tag/v1.8.0)
